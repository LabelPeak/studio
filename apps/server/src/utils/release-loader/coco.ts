import { format } from "date-fns";
import { type ImageClassifyAnnotation } from "shared";

import { BasicReleaseLoader, type DataItem, MAX_ANNOTATION_PER_FILE } from "./basic.ts";

interface CocoInfo {
  description?: string;
  year?: number;
  contributor?: string;
  date_created?: string;
  [key: string]: any;
}

interface CocoImage {
  id: number;
  file_name: string;
  width: number;
  height: number;
  license?: number;
}

interface CocoAnnotation {
  id: number;
  /**
   * Will be matched with CocoImage['id']
   */
  image_id: string | number;
  category_id: number;
  segmentation?: any[];
  area?: number;
  bbox?: [number, number, number, number];
  iscrowd?: 0 | 1;
}

interface CocoCategory {
  id: number;
  name: string;
  supercategory?: string;
}

export interface CocoOutputJSON {
  info: CocoInfo;
  images: CocoImage[];
  annotations: CocoAnnotation[];
  categories: CocoCategory[];
}

export class CocoReleaseLoader extends BasicReleaseLoader {
  public releaseToZip(): Promise<string> {
    return Promise.resolve("");
  }

  public composeCocoOutputJSON(): CocoOutputJSON {
    const presetMap = this.presetLabels.reduce((prev, label, index) => {
      prev.set(label, index + 1);
      return prev;
    }, new Map<string, number>());

    // 1. Compose CocoCategories
    const cocoCategories = this.composeCocoCategories(presetMap);

    const cocoImages: CocoImage[] = [];
    const cocoAnnotations: CocoAnnotation[] = [];

    // 2. Compose CocoImages and CocoAnnotations
    for (let i = 0; i < this.dataItems.length; i++) {
      const dataItem = this.dataItems[i];

      const numericImageId = i + 1;

      // Compose CocoImage
      const image = this.composeCocoImage(dataItem, numericImageId);
      cocoImages.push(image);

      // Compose CocoAnnotation
      const fileAnnotations = this.composeCocoAnnotationByFile(dataItem, numericImageId, presetMap);
      cocoAnnotations.push(...fileAnnotations);
    }

    // 3. Compose CocoInfo
    const cocoInfo: CocoInfo = {
      description: `Dataset: ${this.releaseName} - LabelPeak COCO Export`,
      contributor: this.poc,
      year: new Date().getFullYear(),
      date_created: format(new Date(), "yyyy/MM/dd"),
      version: "1.0"
    };

    return {
      info: cocoInfo,
      images: cocoImages,
      annotations: cocoAnnotations,
      categories: cocoCategories
    };
  }

  private composeCocoCategories(presetMap: Map<string, number>): CocoCategory[] {
    return Array.from(presetMap.entries()).map(([name, id]) => ({
      id,
      name,
      supercategory: "object"
    }));
  }

  private composeCocoImage(dataItem: DataItem, imageId: number): CocoImage {
    const ext = dataItem.file.split(".").pop() || "jpg";

    return {
      id: imageId,
      file_name: `${String(imageId).padStart(10, "0")}.${ext}`,
      width: (dataItem.annotation as ImageClassifyAnnotation[]).at(0)?.originWidth ?? 0,
      height: (dataItem.annotation as ImageClassifyAnnotation[]).at(0)?.originHeight ?? 0
    };
  }

  private composeCocoAnnotationByFile(
    dataItem: DataItem,
    imageId: number,
    presetMap: Map<string, number>
  ): CocoAnnotation[] {
    return (dataItem.annotation as ImageClassifyAnnotation[]).map((annotation, index) => {
      const { originWidth, originHeight } = annotation;
      const { x, y, width, height, labels } = annotation.value;
      const xMin = Math.floor((x * originWidth) / 100);
      const yMin = Math.floor((y * originHeight) / 100);
      const _width = Math.floor((width * originWidth) / 100);
      const _height = Math.floor((height * originHeight) / 100);

      return {
        id: imageId * MAX_ANNOTATION_PER_FILE + index + 1,
        image_id: imageId,
        category_id: presetMap.get(labels[0]) ?? -1,
        bbox: [xMin, yMin, _width, _height],
        area: _width * _height,
        iscrowd: 0
      };
    });
  }
}
