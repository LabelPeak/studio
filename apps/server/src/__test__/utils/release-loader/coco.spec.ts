import { describe, expect, it } from "vitest";

import { CocoReleaseLoader } from "@/utils/release-loader/coco.ts";

import { exampleDataItems, exampleDataset, exampleLabels } from "./input.example.ts";

describe("CocoLoader", () => {
  it("should transform to coco format", () => {
    const releaseName = "<releaseName>";
    const poc = "<poc>";

    const loader = new CocoReleaseLoader({
      releaseName,
      poc,
      dataset: exampleDataset,
      presetLabels: exampleLabels,
      dataItems: exampleDataItems
    });

    const result = loader.composeCocoOutputJSON();

    expect(result).matchSnapshot();
  });
});
