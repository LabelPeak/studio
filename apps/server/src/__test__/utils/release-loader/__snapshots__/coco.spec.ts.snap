// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CocoLoader > should transform to coco format 1`] = `
{
  "annotations": [
    {
      "area": 15314,
      "bbox": [
        105,
        18,
        589,
        26,
      ],
      "category_id": 1,
      "id": 101,
      "image_id": 1,
      "iscrowd": 0,
    },
    {
      "area": 11250,
      "bbox": [
        406,
        207,
        375,
        30,
      ],
      "category_id": 2,
      "id": 102,
      "image_id": 1,
      "iscrowd": 0,
    },
    {
      "area": 7900,
      "bbox": [
        832,
        127,
        395,
        20,
      ],
      "category_id": 4,
      "id": 201,
      "image_id": 2,
      "iscrowd": 0,
    },
    {
      "area": 10720,
      "bbox": [
        414,
        273,
        335,
        32,
      ],
      "category_id": 3,
      "id": 202,
      "image_id": 2,
      "iscrowd": 0,
    },
    {
      "area": 218924,
      "bbox": [
        1593,
        -2,
        478,
        458,
      ],
      "category_id": 4,
      "id": 301,
      "image_id": 3,
      "iscrowd": 0,
    },
    {
      "area": 13790,
      "bbox": [
        232,
        372,
        394,
        35,
      ],
      "category_id": 1,
      "id": 302,
      "image_id": 3,
      "iscrowd": 0,
    },
  ],
  "categories": [
    {
      "id": 1,
      "name": "葡萄",
      "supercategory": "object",
    },
    {
      "id": 2,
      "name": "香蕉",
      "supercategory": "object",
    },
    {
      "id": 3,
      "name": "苹果",
      "supercategory": "object",
    },
    {
      "id": 4,
      "name": "卷心菜",
      "supercategory": "object",
    },
  ],
  "images": [
    {
      "file_name": "0000000001.jpg",
      "height": 992,
      "id": 1,
      "width": 2056,
    },
    {
      "file_name": "0000000002.jpg",
      "height": 992,
      "id": 2,
      "width": 2056,
    },
    {
      "file_name": "0000000003.jpg",
      "height": 992,
      "id": 3,
      "width": 2056,
    },
  ],
  "info": {
    "contributor": "<poc>",
    "date_created": "2025/06/03",
    "description": "Dataset: <releaseName> - LabelPeak COCO Export",
    "version": "1.0",
    "year": 2025,
  },
}
`;
